**For Plotting v_a Against Parameters**



---


**For v_a vs rho (density)**


*   Density is varied and the corresponding value of order parameter v_a is calculated. Since, density is given by


    rho = N/(L*L) (1.1)
    L = 20 (Fixed)
    therefore, N is varied.



*   Noise(eta) is kept constant at some value. 

*   We take 24 different values of N. For all of these different N, the initial random allotment of positions is different. 

*   v_a is calculated for all these 24 different values of rho resulting from changing the number of particles (N) 

*   A plot of rho (x-axis) vs v_a (y-axis) is drawn

**Algorithm** (Aim: Vary rho and calculate v_a)


1.   An array of 24 numbers (all of which denote the number of particles N) is formed
2.   The vicsek simulation is run for this array of numbers (we can either input the number of particles each time or automate it using this array)

3.   v_a is calculated for each of these 24 numbers in the array and all the calculated values are in turned stored in another array v_as (as done before)
4.   rho is calculated using equation (1.1) and the values are stored in an array
5. A plot of rho vs v_a is drawn (with rho in the x-axis and v_a in the y-axis)

---



**For v_a vs eta (noise)**


*   Noise is varied and the corresponding value of order parameter v_a is calculated. 

*   Density (rho) is kept constant at some value. 
*   To keep the density constant, the value of N and L for each run is changed as follows: 
      1. N = 40, L = 3.1
      2. N = 100, L = 5
      3. N = 400, L = 10
      4. N = 4000, L = 31.6
      5. N = 10000, L = 50
*   A value of N and L is taken from the list, and the vicsek simualtion is run several times by changing the noise at each time.

*   v_a is calculated for all these different values of eta resulting from changing the noise (N) 
* The same process is repeated with the other values of N and L from the list

*   A plot of eta (x-axis) vs v_a (y-axis) is drawn for each value of N and L from the list. All of these 5 plots are then merged into a single plot to get a holistic picture

**Algorithm** (Aim: Vary eta and calculate v_a for 5 different values of N and L)

1. The value of N and L is chosen from the list and manually changed into the program

2. An array of 150 numbers is chosen between 0 to 5 (all of which denote eta ) 
2.   For a chosen N and L, the vicsek simulation is run 150 times for this array of numbers containing different values for Noise (eta) 

3.   v_a is calculated for each of these 100 numbers in the array and all the calculated values are in turned stored in another array v_as (as done before)
4.   The process 1, 2, 3 and 4 is then repeated for the remaining values of N and L in the list.
5. A plot of eta vs v_a is drawn (with eta in the x-axis and v_a in the y-axis) for all different values of N and L.
6.  All these plots are merged into one to get a holistic picture

*(Tip: Plot using another platform or software if colab is making it difficult)*

---

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import animation as animation
from IPython import display

In [ ]:
def get_v_a(pos, theta, eta, L, N, t):
  # Define parameters
  r = 1 # Interaction radius of particle
  del_t = 1 # Time step
  v = 0.03 # Velocity of particles

  # Get Vicsek Model final state
  for k in range(t):  
    for i in range(N): 
      x1 = pos[i][0]
      y1 = pos[i][1]
      sine = []
      cosine = []
      # scanning for the ith particle
      for j in range(N):
        x2 = pos[j][0]
        y2 = pos[j][1]
        if (x2-x1)*(x2-x1) + (y2-y1)*(y2-y1) <= r: # scanning the particles present inside and on the boundary of the interaction radius
          sine.append(np.sin(theta[j]))
          cosine.append(np.cos(theta[j]))
      avg_sin = np.average(sine)
      avg_cos = np.average(cosine)
      avg_theta = np.arctan(float(avg_sin/avg_cos)) # average direction of all the particles in the interaction radius
      rand_numb = np.random.uniform(-0.5*eta, 0.5*eta, size = 1) # random noise allotment
      theta[i] = avg_theta + rand_numb[0] # updating theta (direction of velocity of ith particle)

      pos[i][0] = pos[i][0] + v*np.cos(theta[i]) # updating x coordinate of ith particle
      pos[i][1] = pos[i][1] + v*np.sin(theta[i]) # updating y coordinate of ith particle

      # for periodic boundary conditions 
      if pos[i][0] > L:
        pos[i][0] = pos[i][0] - L
      if pos[i][0] < 0:
        pos[i][0] = pos[i][0] + L

      if pos[i][1] > L:
        pos[i][1] = pos[i][1] - L
      if pos[i][1] < 0:
        pos[i][1] = pos[i][1] + L
  
  # Calculate v_a
  c_sum = np.sum(np.cos(theta))
  s_sum = np.sum(np.sin(theta))
  c_square = c_sum*c_sum
  s_square = s_sum*s_sum
  root = np.sqrt(c_square + s_square)
  v_a = root/N
  return v_a

In [ ]:
L = 20 # Length of the cell
eta = 0.7 # Noise
t = 1000 # no. of iterations
N = np.arange(40, 4000, 165) # array of 24 numbers each representing N
rhos = N/(L*L) # Array of particle number densities
v_as = []

# rho versus v_a for fixed noise(eta)
for nu in N: 
  pos = np.random.uniform(0,L,size=(nu,2))
  theta = np.random.uniform(-np.pi, np.pi,size=nu)
  v_a = get_v_a(pos, theta, eta, L, nu, t)
  print(f"Got {v_a} for {nu}.")
  v_as.append(v_a)

fig, ax= plt.subplots(figsize=(6,6))
plt.scatter(rhos,v_as)
plt.xlabel("rho")
plt.ylabel("v_a")
plt.title('rho versus v_a')

In [ ]:
rho = 4 # Particle number density (approx.)
N = 1000
L = 15.8
t = 1000
etas = np.linspace(0.01, 5, 50 ) # Array of noises
v_as = []

pos = np.random.uniform(0,L,size=(N,2))
theta = np.random.uniform(-np.pi, np.pi,size=N)

# eta versus v_a for fixed density
for et in etas:
  v_a = get_v_a(pos, theta, et, L, N, t)
  print(f"Got {v_a} for {et}.")
  v_as.append(v_a)

plt.scatter(etas,v_as)
plt.xlabel("eta")
plt.ylabel("v_a")
plt.title('eta versus v_a')

Got 0.9999931352064491 for 0.01.
Got 0.9988684373223937 for 0.11183673469387755.
Got 0.9955150185357274 for 0.2136734693877551.
Got 0.9915867579806839 for 0.31551020408163266.
Got 0.980812583381166 for 0.4173469387755102.
Got 0.9757312665346594 for 0.5191836734693878.
Got 0.9666055054300585 for 0.6210204081632653.
Got 0.9258740500506797 for 0.7228571428571429.
Got 0.9517748679893917 for 0.8246938775510204.
Got 0.9316409596134902 for 0.926530612244898.
Got 0.9257754066101691 for 1.0283673469387755.
Got 0.9029605389550114 for 1.130204081632653.
Got 0.6482182956680425 for 1.2320408163265306.


\noindent In Figure \ref{fig:2.2}, the change in order parameter $v_a$ is plotted by varying noise and density, keeping the other fixed.\\It can be inferred from Figure \ref{fig:2.2}(a) that as the amount of noise is decreased, the order parameter reaches 1, thus transitioning from disorded to ordered motion.the behaviour of the kinetic order parameter $v_a$ is similar to that of the order parameter of some equilibrium systems (Figure \ref{fig:1.2}) close to their critical point.